_source/_posts/2022-06-28-add-auth-to-any-app-with-oauth2-proxy.adoc

```diff
@@ -0,0 +1,583 @@
+
+In the `docker-compose.yml`, trim down the environment variables to the bare minimum needed for a REST API:
+
+====
+[source,yaml]
+----
+...
+    environment:
+      OAUTH2_PROXY_UPSTREAMS: http://web-app:8080/
+      OAUTH2_PROXY_PROVIDER: oidc # <.>
+      OAUTH2_PROXY_EMAIL_DOMAINS: '*'
+      OAUTH2_PROXY_SKIP_JWT_BEARER_TOKENS: true # <.>
+      OAUTH2_PROXY_OIDC_EMAIL_CLAIM: sub # <.>
+      OAUTH2_PROXY_OIDC_ISSUER_URL: ${ISSUER} # <.>
+      OAUTH2_PROXY_CLIENT_ID: api://default # <.>
+      OAUTH2_PROXY_SET_XAUTHREQUEST: true
+      OAUTH2_PROXY_CLIENT_SECRET: this_value_is_required_but_not_used # <.>
+      OAUTH2_PROXY_COOKIE_SECRET: NOT_USED_BUT_REQUIRED_VALUE_32b_ # <.>
+...
+----
+
+<.> We are not actually using any of the OIDC flows, but this is still required
+<.> Perhaps a poorly named variable, this tells `oauth2-proxy` to validate the JWT access token and to "skip" looking for an OAuth 2.0 session.
+<.> Read the user's email from the `sub` claim in the access token.
+<.> The same issuer URL is used, the JWKS endpoint will be looked up automatically via the OIDC discovery metadata.
+<.> The "client-id" is actually the audience `aud` claim and not the ID of a specific client (multiple API "clients" could be accessing the same REST API).
+<.> There is no "client-secret," but it's a required field...
+<.> Same with the cookie secret, these flows do not use cookies, but the field is required.
+====
+
+Restart the services (stop and then run `docker compose up` again.)
+
+Using the access token environment variable, you set in the previous section run:
+
+[source,bash]
+----
+http :4180/echo "Authorization: Bearer ${TOKEN}"
+----
+
+Awesome! Now your application is secured for REST clients!
+
+Not so fast; now our browser clients don't work correctly, the client ID and secret are not correct, which means users will not be able to log in. We can fix both of these problems with Nginx.
+
+== Add Nginx to route traffic
+
+Adding another reverse proxy into the mix might seem excessive; for a request to get to the application, it would need to pass through both Nginx and OAuth2 Proxy first. However, you may already use Nginx for load balancing, TLS termination, or other ingress concerns.
+
+image::{% asset_path 'blog/add-auth-to-any-app-with-oauth2-proxy/nginx-oauth2-proxy-browser.svg' %}[alt=Sequence Diagram showing a flowing through both request with Nginx and oauth2-proxy,align=center,link={% asset_path 'blog/add-auth-to-any-app-with-oauth2-proxy/nginx-oauth2-proxy-browser.svg' %}]
+
+While we could route our traffic through both proxies as pictured above, I will use the Nginx `auth_request` directive instead. Nginx will make a REST request to OAuth2 Proxy's `/oauth2/auth` endpoint using the original request headers (including any cookies and `Authorization` headers). OAuth2 Proxy will respond with a `202` status code if the request was valid or a `401` otherwise.
```

Comment:
```suggestion
While we could route our traffic through both proxies as pictured above, I will use the Nginx `auth_request` directive instead. Nginx will make a REST request to OAuth2 Proxy's `/oauth2/auth` endpoint using the original request headers (including any cookies and `Authorization` headers). OAuth2 Proxy will respond with a `202` status code if the request is valid or a `401` otherwise.
```
verb tense alignment